crates/wdl-engine/src/cache.rs

```diff
@@ -0,0 +1,635 @@
+            V: Eq,
+        {
+            for (k, v) in a {
+                match b.get(k) {
+                    Some(ov) => {
+                        if v != ov {
+                            bail!("{kind} `{k}` was modified")
+                        }
+                    }
+                    None => bail!("{kind} `{k}` was added"),
+                }
+            }
+
+            for k in b.keys() {
+                if !a.contains_key(k) {
+                    bail!("{kind} `{k}` was removed");
+                }
+            }
+
+            Ok(())
+        }
+
+        if self.command != entry.command {
+            bail!("the command of the task was modified");
+        }
+
+        if self.container != entry.container {
+            bail!("the container used by the task was modified");
+        }
+
+        if self.shell != entry.shell {
+            bail!("the shell used by the task was modified");
+        }
+
+        compare_maps(&self.requirements, &entry.requirements, "task requirement")?;
+        compare_maps(&self.hints, &entry.hints, "task hint")?;
+        compare_maps(&self.inputs, &entry.inputs, "task input")?;
+        Ok(())
+    }
+}
+
+impl fmt::Display for Key {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        self.key.fmt(f)
+    }
+}
+
+/// Represents a request to calculate a [`Key`].
+pub struct KeyRequest<'a> {
+    /// The document containing the task.
+    pub document: &'a Document,
+    /// The task identifier.
+    pub task_id: &'a str,
+    /// The evaluated command of the task.
+    pub command: &'a str,
+    /// The container used by the task.
+    pub container: &'a str,
+    /// The shell used by the task.
+    pub shell: &'a str,
+    /// The evaluated requirements of the task.
+    pub requirements: &'a HashMap<String, Value>,
+    /// The evaluated hints of the task.
+    pub hints: &'a HashMap<String, Value>,
+    /// The backend inputs of the task.
+    pub inputs: &'a [Input],
+}
+
+/// Represents an evaluation call cache.
+///
+/// A call cache can be used to cache the result of task executions so previous
+/// results can be reused and a task's execution skipped.
+///
+/// A [`CallCache`] can be cheaply cloned.
+#[derive(Clone)]
+pub struct CallCache(State);
+
+impl CallCache {
+    /// Creates a new call cache for the given cache directory and file
+    /// transferer to use.
+    ///
+    /// If `cache_dir` is `None`, the default operating system specified cache
+    /// directory for the user is used.
+    pub async fn new(cache_dir: Option<&Path>, transferer: Arc<dyn Transferer>) -> Result<Self> {
+        let cache_dir = match cache_dir {
+            Some(cache_dir) => cache_dir.into(),
+            None => crate::config::cache_dir()?.join(CALL_CACHE_SUBDIR),
+        };
+
+        info!(
+            "using call cache directory `{cache_dir}`",
+            cache_dir = cache_dir.display()
+        );
+
+        fs::create_dir_all(&cache_dir).with_context(|| {
```

Comment:
Could matter on HPCs with slow filesystems to use the `tokio::fs` version instead